api/src/controllers/zot4planimport.ts

```diff
@@ -0,0 +1,173 @@
+/**
+ @module Zot4PlanImportRoute
+*/
+
+import { z } from 'zod';
+import { publicProcedure, router } from '../helpers/trpc';
+import { TRPCError } from '@trpc/server';
+import { SavedRoadmap, SavedPlannerData, SavedPlannerQuarterData, QuarterName } from '@peterportal/types';
+
+type Zot4PlanYears = string[][][];
+
+type Zot4PlanSchedule = {
+  years: Zot4PlanYears;
+  selectedPrograms: {
+    value: number;
+    label: string;
+    is_major: boolean;
+  }[][];
+  addedCourses: [];
+  courses: [];
+  apExam: {
+    id: number;
+    name: string;
+    score: number;
+    courses: string[];
+    GE: [];
+    units: number;
+  }[];
+};
+
+/**
+ * Get a JSON schedule from Zot4Plan by name
+ * Throw an error if it does not exist
+ */
+const getFromZot4Plan = async (scheduleName: string) => {
+  let res = {};
+  await fetch('https://api.zot4plan.com/api/loadSchedule/' + scheduleName, {
+    method: 'PUT',
+  })
+    .then((response) => {
+      if (!response.ok) {
+        throw new TRPCError({
+          code: 'BAD_REQUEST',
+          message: 'Schedule name could not be obtained from Zot4Plan',
+        });
+      }
+      return response.json();
+    })
+    .then((json) => {
+      res = json;
+    });
+  return res as Zot4PlanSchedule;
+};
+
+/**
+ * Convert a Zot4Plan course name into a PeterPortal course ID
+ */
+const convertIntoCourseID = (zot4PlanCourse: string): string => {
+  // PeterPortal course IDs are the same as Zot4Plan course IDs except all spaces are removed
+  return zot4PlanCourse.replace(/\s/g, '');
+};
+
+/**
+ * Trim the empty years off the end of a saved roadmap planner
+ * (Other than the first year)
+ */
+const trimEmptyYears = (planner: SavedPlannerData) => {
+  // Empty years in the middle aren't trimmed because that makes it hard to add years there
+  while (planner.content.length > 1) {
+    let yearHasCourses = false;
+    for (const quarter of planner.content[planner.content.length - 1].quarters) {
+      if (quarter.courses.length != 0) {
+        yearHasCourses = true;
+      }
+    }
+    if (!yearHasCourses) {
+      // The year does not have courses, so trim it
+      planner.content.pop();
+    } else {
+      // The year does have courses, so we are done
+      break;
+    }
+  }
+};
+
+/**
+ * Determine a student's start year based on their current year in school
+ * (ex. a first-year's current year is "1")
+ */
+const getStartYear = (studentYear: string): number => {
+  let startYear = new Date().getFullYear();
+  startYear -= parseInt(studentYear);
+  // First-years in Fall start this year, not the previous year
+  if (new Date().getMonth() >= 7) startYear += 1;
```

Comment:
Technically this should work (might want to specify via comment that this is August, not July), but in a perfect world this would get schedules from [the API](https://anteaterapi.com/reference#tag/calendar/GET/v2/rest/calendar/all) and use that to determine what academic year we're in.

This can also be a separate issue since that's sorta separate from the rest of imports. cc @ecxyzzy for thoughts